src/zarr/abc/store.py

```diff
@@ -386,6 +388,32 @@ async def _get_many(
         for req in requests:
             yield (req[0], await self.get(*req))
 
+    async def getsize(self, key: str) -> int:
+        """
+        Return the size, in bytes, of a value in a Store.
+
+        Parameters
+        ----------
+        key : str
+
+        Returns
+        -------
+        nbytes: int
+            The size of the value in bytes.
+
+        Raises
+        ------
+        FileNotFoundError
+            When the given key does not exist in the store.
+        """
+        # Note to implementers: this default implementation is very inefficient since
+        # it requires reading the entire object. Many systems will have ways to get the
+        # size of an object without reading it.
+        value = await self.get(key, prototype=default_buffer_prototype())
```

Comment:
I think there's a good reason to avoid dynamic default arguments, e.g. function invocations (and I think there's a linting rule that complains about such things). But that doesn't change the fact that `default_buffer_prototype()` _is_ the default, which makes it annoying to specify. One idea that came up in the past was to statically associate a store with a buffer prototype, e.g. by making the store class generic, but this would reduce the flexibility of the store... although whether we realistically need the same store instance to return multiple different types of buffers is yet to be seen.

Reply:
Would you be OK with `defualt_buffer_prototype` being an optional argument (`| None`, with a default of None, which means use `default_buffer_prototype()`?

Reply:
yeah, some value like `None` or the literal string `"default"` would be perfect for this, at least in my opinion.

Reply:
Two small bits of awkwardness:

1. We'll now require Store implementors to *read the documentation* to understand what the value of `None / "default"` for `prototype` means when they're implementing `get`. I think that's fine, but merits documenting.
2. I might have missed it, but IIUC `zarr.core` is private, but `zarr.core.buffer` isn't re-exported from anywhere else. I'll add it to `zarr.buffer` unless anyone objects.

Reply:
> One idea that came up in the past was to statically associate a store with a buffer prototype

This thread came up in #1661, where I was struggling to understand the purpose of the prototype argument.

I think it would be nice to either statically associate stores with buffer prototypes or remove the prototypes altogether and let a store return a protocol instead of a specific class.

Reply:
I've removed the buffer related changes from this PR (xref https://github.com/zarr-developers/zarr-python/issues/2473, which might bear on this parameter).